pkg/kubelet/volume/empty_dir/empty_dir.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package empty_dir
+
+import (
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+func init() {
+	volume.RegisterPlugin(&emptyDirPlugin{nil, false})
```

Comment:
An example would be when we want to add a command line flag that prevents a volume plugin from being used (because it's insecure, out of policy, requires something not installed, etc).  You'd want to be able to disable the registered plugin, but it would be long after init().  Of if you want to pass command line arguments to a plugin, and do it on initialization so you don't have to expose public setters.

Static initializers are creeper vines - they send up little shoots everywhere spreading static assumptions.  When it comes time to rip them out, the longer you've wait the longer they've spread.  That's my personal experience from geard - I went down this exact path, started to need registration with ordering (I want this extension before that, so I'll register this plugin with magic number 4), needed flags and conditional load logic, realized I was backed into a corner, and had to rewrite it all.  

> On Jan 18, 2015, at 12:26 AM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/kubelet/volume/empty_dir/empty_dir.go:
> 
> > +
> > +package empty_dir
> > +
> > +import (
> > +	"fmt"
> > +	"os"
> > +	"path"
> > +
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
> > +)
> > +
> > +func init() {
> > +	volume.RegisterPlugin(&emptyDirPlugin{nil, false})
> The flip side to this is that I can not have drop-in exec-based plugins without them self-registering.
> 
> I could maybe go halfway - explicitly call InitPlugins() on each plugin pkg, and have that call host.RegisterPlugin() or something.  Will think on it.
> 
> —
> Reply to this email directly or view it on GitHub.
>